### PR TITLE
Add Ethereum Signature Verification

### LINKED ISSUE #30: Try to verify an Ethereum signature match EIP712 hash on StarkNet

### PR DESCRIPTION
**Pull request by pscott**
This PR:
- Adds the `EthSig.cairo` authenticator which verifies a classic Ethereum signature (`r, s, v`) over an EIP-712 hash (hashes types are in `tests/starknet/helper/types.ts`).
- Modifies `felt_to_uint256` to use `split_felt` from the `common` library

Note: `metadataURI` is NOT added to the `Propose` type because `strings` are a little bit hash in Cairo. This should come in a later update :) (tracking issue: https://github.com/snapshot-labs/sx-core/issues/202)

Closes #30 !! 🥳

### CODE REVIEW

**Comment by pscott** [comment]
@Orland0x wtf test is failing in the StorageProof side of things? Any idea of what's going on here? I guess it has to do with a cairo version maybe?


https://github.com/snapshot-labs/sx-core/runs/7044640255?check_suite_focus=true

**Comment by pscott** [comment]
@Orland0x wtf test is failing in the StorageProof side of things? Any idea of what's going on here? 
https://github.com/snapshot-labs/sx-core/runs/7044640255?check_suite_focus=true

**Comment by Orland0x** [comment]
> 

Hmmm `0x3ae0...` is the selector for `get_latest_l1_block` in Fossil so I think its to do with the an old version of Fossil (that didnt have that method) getting deployed. 
try adding `git submodule update --remote` to the GA

File: README.md [hunk 1]
```diff
@@ -28,6 +28,7 @@ contracts
 │   │  ├─ array2d.cairo — "For handling 2 dimensional arrays"
 │   │  ├─ choice.cairo — "The set of choices one can make for a vote"
 │   │  ├─ eth_address.cairo — "Ethereum address type"
+│   │  ├─ felto_to_uint256.cairo — "Generate a uint256 from a felt"
```
Comment (Orland0x) [inline README.md:32]: typo, also need to add `EthSig.cairo`

File: contracts/starknet/lib/felt_to_uint256.cairo [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from starkware.cairo.common.cairo_builtins import BitwiseBuiltin
```
Comment (Orland0x) [inline contracts/starknet/lib/felt_to_uint256.cairo]: redundant imports lines 1 and 3

File: contracts/starknet/Authenticators/EthSig.cairo [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        # Divide by 16
+        let (divided, _) = uint256_unsigned_div_rem(num, Uint256(16, 0))
+
+        let (res_len) = get_base16_len(divided)
+        return (res_len + 1)
+    end
+end
+
+# Computes `base ** exp` where `base` and `exp` are both `felts` and returns the result as a `Uint256`.
+func u256_pow{range_check_ptr}(base : felt, exp : felt) -> (res : Uint256):
```
Comment (Orland0x) [inline contracts/starknet/Authenticators/EthSig.cairo:131]: any reason to not use [pow](https://github.com/starkware-libs/cairo-lang/blob/167b28bcd940fd25ea3816204fa882a0b0a49603/src/starkware/cairo/common/pow.cairo) from standard lib then just convert to Uint256 ?

File: contracts/starknet/Authenticators/EthSig.cairo [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+    # Ensure proposer has not already used this salt in a previous action
+    let (already_used) = salts.read(proposer_address, salt)
+
+    with_attr error_message("Salt already used"):
+        assert already_used = 0
+    end
+
+    # Execution_hash should be in calldata[1] and calldata[2]
+    let execution_hash = Uint256(calldata[1], calldata[2])
```
Comment (Orland0x) [inline contracts/starknet/Authenticators/EthSig.cairo]: We now dont store a separate execution hash, check this PR: https://github.com/snapshot-labs/sx-core/pull/194

File: test/shared/setup.ts [hunk 5]
```diff
@@ -573,3 +573,64 @@ export async function starkTxAuthSetup() {
...
+
+  const vanillaSpaceFactory = await starknet.getContractFactory('./contracts/starknet/Space.cairo');
+  const vanillaVotingStategyFactory = await starknet.getContractFactory(
+    './contracts/starknet/VotingStrategies/Vanilla.cairo'
+  );
+  const starknetSigAuthFactory = await starknet.getContractFactory(
+    './contracts/starknet/Authenticators/EthSig.cairo'
+  );
+  const zodiacRelayerFactory = await starknet.getContractFactory(
+    './contracts/starknet/ExecutionStrategies/ZodiacRelayer.cairo'
```
Comment (Orland0x) [inline test/shared/setup.ts]: I was generally using vanilla contracts for the modules that were not part of the current test, just to allow more isolation of the aspect under test and also reduce GA runtime (as the vanilla contracts are super light).

File: test/shared/setup.ts [hunk 6]
```diff
@@ -573,3 +573,64 @@ export async function starkTxAuthSetup() {
...
 }
+
+export async function ethereumSigSetup() {
+  const account = await starknet.deployAccount('OpenZeppelin');
+
+  const vanillaSpaceFactory = await starknet.getContractFactory('./contracts/starknet/Space.cairo');
+  const vanillaVotingStategyFactory = await starknet.getContractFactory(
+    './contracts/starknet/VotingStrategies/Vanilla.cairo'
+  );
+  const starknetSigAuthFactory = await starknet.getContractFactory(
```
Comment (Orland0x) [inline test/shared/setup.ts]: `ethSigAuthFactory`

File: test/starknet/EthereumSigAuth.test.ts [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+  if (sig.startsWith('0x')) {
+    sig = sig.substring(2);
+  }
+  const r = SplitUint256.fromHex('0x' + sig.substring(0, 64));
+  const s = SplitUint256.fromHex('0x' + sig.substring(64, 64 * 2));
+  const v = BigInt('0x' + sig.substring(64 * 2));
+  return { r, s, v };
+}
+
+describe('Ethereum Sig Auth testing', () => {
```
Comment (Orland0x) [inline test/starknet/EthereumSigAuth.test.ts:72]: I refactored all the tests to make them consistent and more readable. Would be good to adapt this to that, check `Space.test.ts` for inspo

File: README.md (hunk 1 above)
Comment (pscott) [reply README.md:32]: duh haha

File: contracts/starknet/lib/felt_to_uint256.cairo (hunk 2 above)
Comment (pscott) [reply contracts/starknet/lib/felt_to_uint256.cairo]: makes me think about https://github.com/snapshot-labs/sx-core/issues/97 :)

File: contracts/starknet/Authenticators/EthSig.cairo (hunk 4 above)
Comment (pscott) [reply contracts/starknet/Authenticators/EthSig.cairo]: ok let's merge #194 first and then I'll update this!

File: contracts/starknet/Authenticators/EthSig.cairo (hunk 3 above)
Comment (pscott) [reply contracts/starknet/Authenticators/EthSig.cairo:131]: Hehe yep I first used `pow` and spent some time investigating why my hash wasn't working :)
Indeed `felt` is < `Uint256`, so it overflows and just returns an invalid value :)

Comment (pscott) [reply contracts/starknet/Authenticators/EthSig.cairo:131]: (and indeed `pow` doesn't throw when an overflow happens... learned it the hard way :p)

File: test/shared/setup.ts (hunk 5 above)
Comment (pscott) [reply test/shared/setup.ts]: definitely right

File: test/starknet/EthereumSigAuth.test.ts (hunk 7 above)
Comment (pscott) [reply test/starknet/EthereumSigAuth.test.ts:72]: done :D

File: contracts/starknet/Authenticators/EthSig.cairo (hunk 3 above)
Comment (Orland0x) [reply contracts/starknet/Authenticators/EthSig.cairo:131]: ah i see interesting

File: contracts/starknet/Authenticators/EthSig.cairo [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+const DOMAIN_HASH_HIGH = 0x4ea062c13aa1ccc0dde3383926ef9137
+const DOMAIN_HASH_LOW = 0x72c5ab51b06b74e448d6b02ce79ba93c
+
+# Maps a tuple of (user, salt) to a boolean stating whether this tuple was already used or not (to prevent replay attack).
+@storage_var
+func salts(user : felt, salt : Uint256) -> (already_used : felt):
+end
+
+# Adds a 2 bytes (16 bits) `prefix` to a 16 bytes (128 bits) `value`.
+func add_prefix128{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(value : felt, prefix : felt) -> (
```
Comment (Orland0x) [inline contracts/starknet/Authenticators/EthSig.cairo:58]: might be worth moving these utilities to the `lib` folder? Just leaving `authenticate`, `authenticate_vote` and `authenticate_proposal` in this file

File: test/starknet/helper/types.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import { number } from 'starknet';
```
Comment (Orland0x) [inline test/starknet/helper/types.ts]: would make sense to put this `types.ts` file in `test/shared` rather than here i think

**Review by Orland0x — APPROVED** [review]

**Pull request merged**